### DISCUSSION TITLE
Deploying to production after git push

### DISCUSSION THREAD

**Discussion by leo-petrucci** [community/community#22123] [New to GitHub]
<p>I’m currently trying to optimize my workflow but I’m a bit lost on how to cut down deployment time once I push to production.</p>
<p>At the moment my workflow is something like:</p>
<ol>
<li>Commit to branch</li>
<li>Merge to master</li>
<li>Wait for tests to complete</li>
<li>Login into my production server</li>
<li>Manually pull and build</li>
</ol>
<p>While it’s not prone to many errors, it’s definitely an annoyance to do over and over since the process is always the same.</p>
<p>I’m wondering if there’s any way to automatically pull changes and build my application once I push changes to my Github repository and the tests successfully pass?</p>
<p>I’m aware of Jenkins, but it seems like it would add a lot of overhead that I don’t really need since most of my repositories are simple websites.</p>
<p>Is there a way to do this without external tools? Ideally I’m looking to implement something that would:</p>
<ol>
<li>Recognize when tests are passed on github</li>
<li>Pull changed code</li>
<li>Install dependencies and build the application</li>
<li>Reload the existing application</li>
</ol>
<p><strong>Note:</strong> I’ve found a lot of tutorials that involve creating a bare repo on the server and then manually pushing code to that. This wouldn’t be a great solution as I would essentially need to push to two different repositories every time. But if there’s any way to implement this with Github being the middle man that would be fantastic!</p>

**Comment by airtower-luna** [comment]
<p>This is something you could do with Github Actions. Depending on how much access you want to give Actions to your production system I see two concepts you might use:</p>
<ul>
<li>Set up a script for pull and build on your production server, and some way to trigger it from Actions (e.g. send a HTTP POST request to a certain URL). I’d recommend using some sort of authentication token so random people can’t trigger builds, but otherwise it’s nicely encapsulated on your server.</li>
<li>Integrate the deploy script into an Actions workflow, and set up SSH access or similar using actions secrets so the workflow can log in to your server and run the deploy.</li>
</ul>
<p>Personally I’d lean more towards the first because there’s less risk of production server credentials leaking. As an extra precaution you could sign commits (or tags) and verify them on your production server before building. :slightly_smiling_face:</p>

**Comment by leo-petrucci** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="2" data-topic="120898">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>Set up a script for pull and build on your production server, and some way to trigger it from Actions (e.g. send a HTTP POST request to a certain URL). I’d recommend using some sort of authentication token so random people can’t trigger builds, but otherwise it’s nicely encapsulated on your server.</p>
</blockquote>
</aside>
<p>How would this work exactly? Is there a way to hide auth tokens within Github Actions?</p>
<aside class="quote no-group" data-username="airtower-luna" data-post="2" data-topic="120898">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>Integrate the deploy script into an Actions workflow, and set up SSH access or similar using actions secrets so the workflow can log in to your server and run the deploy.</p>
</blockquote>
</aside>
<p>Same question for this.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="creativiii" data-post="3" data-topic="120898">
<div class="title">
<div class="quote-controls"></div>
 creativiii:</div>
<blockquote>
<p>Is there a way to hide auth tokens within Github Actions?</p>
</blockquote>
</aside>
<p>Yes, it’s called Secrets. Take a look at the documentation: <a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets" rel="nofollow noopener"> Creating and storing encrypted secrets</a></p>

**Comment by leo-petrucci** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="4" data-topic="120898">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>Yes, it’s called Secrets. Take a look at the documentation: <a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets" rel="nofollow noopener"> Creating and storing encrypted secrets</a></p>
</blockquote>
</aside>
<p>Oh that’s great! No idea that was a thing.</p>
<p>So I can use something like <a href="https://github.com/marketplace/actions/http-request-action" rel="nofollow noopener">this</a> to make a request to a PHP file that runs shell_exec functions?</p>
<p>Is there a specific way to ensure the HTTP request gets sent <strong>only if</strong> the previous tests pass? Do I just need to add it after my tests?</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="creativiii" data-post="5" data-topic="120898">
<div class="title">
<div class="quote-controls"></div>
 creativiii:</div>
<blockquote>
<p>So I can use something like <a href="https://github.com/marketplace/actions/http-request-action" rel="nofollow noopener">this</a> to make a request to a PHP file that runs shell_exec functions?</p>
</blockquote>
</aside>
<p>I’d probably just use <code>curl</code>, but sure!</p>
<aside class="quote no-group" data-username="creativiii" data-post="5" data-topic="120898">
<div class="title">
<div class="quote-controls"></div>
 creativiii:</div>
<blockquote>
<p>Is there a specific way to ensure the HTTP request gets sent <strong>only if</strong> the previous tests pass? Do I just need to add it after my tests?</p>
</blockquote>
</aside>
<p>If your test step returns a non-zero exit code so the Actions runner will consider it failed, yes. A step that doesn’t <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions" rel="nofollow noopener">set another <code>if:</code> condition</a> in the step definition won’t run after a previous failed step.</p>
<p>If the test exit status doesn’t work I’d suggest fixing that, but you could also check for it some other way and write your own <code>if:</code> condition for the deploy step.</p>

**Comment by leo-petrucci** [comment]
<p>Amazing! Thank you for all the help, I’ll try implementing this over the weekend :slight_smile:</p>

**Comment by leo-petrucci** [comment]
<p>I’m almost done but I can’t seem to be able to run a curl command from my Github Actions:</p>
<pre><code>  - name: The job has succeeded
    with:
      token: ${{ secrets.SYNC_SCRIPT_TOKEN }}
    if: ${{ success() }}
    run: curl -X GET http://creativiii.com/deploy.php -H 'Authorization: $token' 
</code></pre>
<p>Github is telling me " <code>You have an error in your yaml syntax on line 29</code>" referring to</p>
<pre><code class="lang-auto">run: curl -X GET http://creativiii.com/deploy.php -H 'Authorization: $token'
</code></pre>
<p>This works on my local machine so I’m not sure why it wouldn’t here.<br>
Any help would be greatly appreciated.</p>

**Comment by airtower-luna** [comment]
<p>First off, I’ve recently seen a report of off-by-one errors on syntax error reporting in Github Actions. Not sure if that’s been fixed yet, so I’d recommend looking at lines 28 (looks good to me, though redundant) and 30 as well.</p>
<p>A thing you could try to make sure that the quotes aren’t confusing the YAML parser would be to use a <a href="https://yaml-multiline.info/" rel="nofollow noopener">YAML block string</a>:</p>
<pre><code class="lang-yaml">    run: |
      curl -X GET http://creativiii.com/deploy.php -H 'Authorization: $token'
</code></pre>
<p>However I see two other issues:</p>
<ol>
<li>To use <code>token</code> as an environment variable in your command you need to set it using an <code>env</code> block, not <code>with</code>:</li>
</ol>
<pre><code class="lang-yaml">    env:
      token: ${{ secrets.SYNC_SCRIPT_TOKEN }}
</code></pre>
<ol start="2">
<li>Don’t use a GET request for something that changes state on the server. In HTTP a <a href="https://tools.ietf.org/html/rfc7231.html#section-4.3.1" rel="nofollow noopener">GET request is meant to retrieve information</a> from the server, not trigger additional action. POST would be an appropriate method to use. I made the same error a few years back, fresh out of university, please don’t repeat it. :wink:
</li>
</ol>

**Comment by facundofarias** [comment]
Maybe you should give [DeployHQ](https://www.deployhq.com/) a try! 😄

**Reply by leo-petrucci** [reply]
> Maybe you should give *spammy link* a try! 😄 (spammy link removed by litebrite82)

Really? On an 4 year old thread?
Jesus man
